datashader/bundling.py

```diff
@@ -0,0 +1,266 @@
+
+
+@ngjit
+def advect_segments(segments, vert, horiz):
+    for i in range(1, len(segments) - 1):
+        x = int(segments[i][0] * ACCURACY)
+        y = int(segments[i][1] * ACCURACY)
+        segments[i][0] = segments[i][0] + horiz[x, y] / ACCURACY
+        segments[i][1] = segments[i][1] + vert[x, y] / ACCURACY
+        segments[i][0] = max(0, min(segments[i][0], 1))
+        segments[i][1] = max(0, min(segments[i][1], 1))
+
+
+def advect_and_resample(vert, horiz, segments, iterations=50):
+    for it in range(iterations):
+        advect_segments(segments, vert, horiz)
+        if it % 2 == 0:
+            segments = resample_edge(segments)
+    return segments
+
+
+@delayed
+def advect_resample_all(gradients, edge_segments):
+    vert, horiz = gradients
+    return [advect_and_resample(vert, horiz, edges) for edges in edge_segments]
+
+
+def batches(l, n):
+    """Yield successive n-sized batches from l."""
+    for i in range(0, len(l), n):
+        yield l[i:i + n]
+
+
+@delayed
+def draw_to_surface(edge_segments, bandwidth):
+    img = np.zeros((ACCURACY + 1, ACCURACY + 1))
+    for segments in edge_segments:
+        for point in segments:
+            img[int(point[0] * ACCURACY), int(point[1] * ACCURACY)] += 1
+    return gaussian(img, sigma=bandwidth / 2)
+
+
+@delayed
+def get_gradients(img):
+    img /= np.max(img)
+
+    horiz = sobel_h(img)
+    vert = sobel_v(img)
+
+    magnitude = np.sqrt(horiz**2 + vert**2) + 1e-5
+    vert /= magnitude
+    horiz /= magnitude
+    return (vert, horiz)
+
+
+def _from_pandas(nodes, edges):
+    def minmax_scale(series):
+        minimum, maximum = np.min(series), np.max(series)
+        return (series - minimum) / (maximum - minimum)
+
+    nodes['x'] = minmax_scale(nodes['x'])
+    nodes['y'] = minmax_scale(nodes['y'])
+
+    df = pd.merge(nodes, edges, left_index=True, right_on=['source'])
+    df = df.rename(columns={'x': 'src_x', 'y': 'src_y'})
+
+    df = pd.merge(nodes, df, left_index=True, right_on=['target'])
+    df = df.rename(columns={'x': 'dst_x', 'y': 'dst_y'})
+
+    df = df.filter(items=['src_x', 'src_y', 'dst_x', 'dst_y'])
+
+    edge_segments = []
+    for edge in df.get_values():
+        segments = [[edge[0], edge[1]], [edge[2], edge[3]]]
+        edge_segments.append(np.array(segments))
+    return edge_segments
+
+
+def _to_pandas(edge_segments):
+    # Need to put a [np.nan, np.nan] between edges
+    def edge_iterator():
+        for edge in edge_segments:
+            yield edge
+            yield np.array([[np.nan, np.nan]])
+
+    df = DataFrame(np.concatenate(list(edge_iterator())))
+    df.columns = ['x', 'y']
+    return df
+
+
+def nop_bundle(nodes, edges, initial_bandwidth=0.05, decay=0.7, iterations=4, batch_size=20000):
+    # Import from Pandas DataFrame
+    edges = _from_pandas(nodes, edges)
+
+    # Convert to Pandas DataFrame
+    return _to_pandas(edges)
+
+
+def bundle(nodes, edges, initial_bandwidth=0.05, decay=0.7, iterations=4, batch_size=20000):
```

Comment:
This algorithm is just one bundling approach, out of many possible, and so it would be good to have a specific name.  Maybe hammer_bundle?